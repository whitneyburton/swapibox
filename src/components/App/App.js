import React, { Component } from 'react';
import '../../main.scss';
import { fetchData } from '../../helpers/apiCalls';
import * as api from '../../helpers/apiHelpers';
import { Header } from '../Header/Header'
import { FilmScript } from '../FilmScript/FilmScript'
import { Controls } from '../Controls/Controls'
import { CardContainer } from '../CardContainer/CardContainer'

class App extends Component {
  constructor() {
    super();
    this.state = {
      error: null,
      planets: [],
      people: [],
      vehicles: [],
      filmscript: '',
      category: '',
      favorites: []
    }
  };

  componentDidMount = () => {
    this.generateFilmScript();
  }

  generateFilmScript = async () => {
    try {
      const data = await fetchData('https://swapi.co/api/films/');
      const randomNumber = Math.floor(Math.random() * Math.floor(7));
      let featureFilmScript = data.results[randomNumber];
      this.setState({
        filmscript: {
          title: featureFilmScript.title,
          opening_crawl: featureFilmScript.opening_crawl,
          release_date: featureFilmScript.release_date
        }
      })
    } catch (error) {
      this.setState({ error })
    }
  };

  generateVehicles = async () => {
    try {
      const vehicles = await api.fetchVehicles();
      this.setState({ vehicles })
    } catch (error) {
      this.setState({ error })
    }
  };

  generatePlanets = async () => {
    try {
      const planets = await api.fetchPlanets();
      this.setState({ planets })
    } catch (error) {
      this.setState({ error })
    }
  };

  generatePeople = async () => {
    try {
      const people = await api.fetchPeople();
      this.setState({ people })
    } catch (error) {
      this.setState({ error })
    }
  };

  handleFavorite = (card) => {
    let { favorites } = this.state;
    let newFavorites;
    if (favorites.length === 0) {
      newFavorites = [card];
    } else if (favorites.includes(card)) {
      newFavorites = favorites.filter(favorite => {
        return favorite !== card;
      });
    } else {
      newFavorites = [...favorites, card];
    }
    this.setState({ favorites: newFavorites })
  }

  retrieveCategory = (category) => {
    if (category === 'people' && this.state.people.length === 0) {
      this.setState({ category })
      this.generatePeople();
    } else if (category === 'people') {
      this.setState({ category })
    } else if (category === 'planets' && this.state.planets.length === 0) {
      this.setState({ category })
      this.generatePlanets();
    } else if (category === 'planets') {
      this.setState({ category })
    } else if (category === 'vehicles' && this.state.vehicles.length === 0) {
      this.setState({ category })
      this.generateVehicles();
    } else if (category === 'vehicles') {
      this.setState({ category })
    } else {
      this.setState({ category: 'favorites' })
    }
  };

  render() {
    let { category, favorites, filmscript } = this.state;
    let { retrieveCategory, handleFavorite } = this;
    let body;

    if (category) {
      body = <CardContainer
        category={category}
        cards={this.state[category]}
        favoriteCards={favorites}
        handleFavorite={handleFavorite} />
    } else {
      body = <FilmScript
        filmscript={filmscript} />
    }

    return (
      <div className="App">
        <Header />
        <Controls
          favorites={favorites}
          retrieveCategory={retrieveCategory} />
        {body}
      </div>
    );
  }
}

export default App;