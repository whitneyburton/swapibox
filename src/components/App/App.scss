body {
  background-image: url('./images/black-space.jpg');
  background-repeat: repeat;
  background-size: cover;
  background-color: #000;
}

.App {
  font-family: Arial, Geneva, Tahoma, sans-serif;
  // display: grid;
  grid-template-columns: 26% 37% 37%;
  grid-template-rows: 1fr 1fr 8fr;
  grid-template-areas: 
    "header header header"
    "controls controls controls"
    "filmscript cardcontainer cardcontainer";
}