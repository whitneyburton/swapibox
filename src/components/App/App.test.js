import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />)
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  
})
