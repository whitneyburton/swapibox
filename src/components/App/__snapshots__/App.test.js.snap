// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Header />
  <Controls
    retrieveCategory={[Function]}
  />
  <FilmScript
    filmscript=""
  />
  <CardContainer
    category=""
    people={Array []}
    planets={Array []}
    vehicles={Array []}
  />
</div>
`;
